### PR TITLE
feature: read-only copies of user repo's submodules

### LINKED ISSUE #161: [Bug]: Does not work with git submodules
### What happened?

```
⏺ container-use:environment_create (MCP)(title: "Add version extraction from image property", environment_source:
                                        "/Users/gerhard/github.com/dagger/dagger.io-worktree/api-parc-use-cli-version-in-cache-volume", explanation:
                                        "Creating environment to implement version extraction from enginespec image property")
  ⎿  Error: failed to create environment: failed to commit worktree changes: git command failed (exit code 128): exit status 128
     Output: fatal: not a git repository: dagviz/dagger/../../.git/modules/dagviz/dagger
```

As discovered with @aluzzardi 

Image

### Version

```shell
cu version
cu version 0.1.1
commit: b1f20ab328c1e6b33d55cfc332607e5de45bbbaa
built: 2025-06-30T17:35:03Z
```

### Logs

```text
We don't need this. The template should not make this a required value.
```

### LINKED ISSUE #272: [Bug]: SSH_AUTH_SOCK is possibly not being passed to Dagger sessions
### What happened?

When working with an AI agent on a local Git repo that has a submodule pointing to a remote repository that requires SSH authentication, the resulting Dagger session does not have access to the host's SSH socket.

It appears [here](https://github.com/dagger/container-use/blob/71346865565c2a0d9dd92350ed30c86ed73e43b9/repository/repository.go#L203-L212) is where Dagger is invoked during the above, and I think during the Tree() call is when something tries to initialized the submodules. During this process the ssh agent socket at `SSH_AUTH_SOCK` on the host has the required key loaded. But the environment creation fails when git tries to do an SSH call to the remote and can't find any credentials.

Originally, I thought the issue was on Dagger's side, but after opening [an issue there](https://github.com/dagger/dagger/issues/10854), I don't think that's the case. I can't find anywhere container-use code accesses the host's `SSH_AUTH_SOCK` environment variable or configures the Dagger session to use it when it invokes Dagger.

This may be related to #161 but the specifics of our reports are quite different, so I felt it worth posting. I'm willing to dig into the code more and contribute an MR or help get an existing one to the finish line if that existing submodules branch is still valid and worth forking.

### Version

```shell
container-use version 0.4.1
commit: 71346865565c2a0d9dd92350ed30c86ed73e43b9
built: 2025-08-01T18:32:52Z
```

### Logs

```text

```

### PR DESCRIPTION
**Pull request by cwlbraa**
This PR adds very basic support for repositories containing submodules. Submodules in the user's repo are initialized inside the container, but we best-effort block additional writes and commits to the submodules content.

fixes #161
fixes #272 

TODO:
- [x] test that recursive submodules work (in theory, they do, but wanna check)
- [x] test that run_cmd can, in fact, work around restrictions to writing submodule files, and assert what happens if it does (i think right now it will avoid committing those changes or moving the submodule pointer, but it's hard to say)
   - this is not the ideal behavior, but it's good enough. To get better behavior we'd need to get very clever about file permissions or seccomp, and that can come at a later date. the test will include a comment describing how the behavior is less than ideal)

8/19: ready to merge

### CODE REVIEW

File: environment/filesystem.go [hunk 1]
```diff
@@ -183,3 +199,80 @@ func getMatchContext(contents string, matchIndex int) string {
...
+	if err != nil {
+		// If .gitmodules doesn't exist, there are no submodules
+		return submodulePaths
+	}
+
+	// Execute git submodule foreach to get submodule paths
+	// We use a simple approach: run the command and parse the output
+	output, err := env.container().
+		WithWorkdir(env.State.Config.Workdir).
+		WithExec([]string{"sh", "-c", "git submodule foreach --recursive --quiet 'echo $sm_path' 2>/dev/null || true"}).
```
Comment (DaggerAssistant) [inline environment/filesystem.go]: 🐛 **Bug (High):** **Shell injection vulnerability**: The git command construction using shell expansion `'echo $sm_path'` is vulnerable to injection if submodule paths contain malicious characters. This could allow arbitrary command execution.

**Evidence**: The command `git submodule foreach --recursive --quiet 'echo $sm_path'` directly interpolates the `$sm_path` variable without escaping, and malicious submodule paths could inject shell commands.

**Impact**: Potential arbitrary code execution in container context.

```suggestion
		WithExec([]string{"git", "submodule", "foreach", "--recursive", "--quiet", "echo $sm_path"}).
```

File: repository/git.go [hunk 2]
```diff
@@ -407,12 +427,61 @@ func (r *Repository) commitWorktreeChanges(ctx context.Context, worktreePath, ex
...
+
+	// Use git submodule foreach to get submodule paths
+	output, err := RunGitCommand(ctx, worktreePath, "submodule", "foreach", "--recursive", "--quiet", "echo $sm_path")
+	if err != nil {
+		// If command fails, return empty slice - don't block regular operation
+		slog.Debug("Failed to get submodule paths", "error", err)
+		return submodulePaths
+	}
+
+	for line := range strings.SplitSeq(strings.TrimSpace(output), "\n") {
```
Comment (DaggerAssistant) [inline repository/git.go:544]: 🔧 **Maintainability (Medium):** **Inconsistent submodule detection logic**: Two different implementations exist for detecting submodule paths - one in `environment/filesystem.go:detectSubmodulePaths()` and another in `repository/git.go:getSubmodulePaths()`. They use different approaches:

**repository/git.go** (line 448): Uses `strings.SplitSeq()` with range iteration
**environment/filesystem.go** (line 227): Uses `strings.Split()` with index iteration

**Risk**: These subtle differences could lead to inconsistent behavior between host and container submodule detection.

```suggestion
	for _, line := range strings.Split(strings.TrimSpace(output), "\n") {
```

File: environment/integration/helpers.go [hunk 3]
```diff
@@ -131,6 +131,12 @@ var (
 		writeFile(t, repoDir, "README.md", "# Test Project\n")
 		gitCommit(t, repoDir, "Initial commit")
 	}
+
+	SetupRepoWithSubmodule = func(t *testing.T, repoDir string) {
```
Comment (DaggerAssistant) [inline environment/integration/helpers.go]: 🔧 **Maintainability (Low):** **Unused test helper**: `SetupRepoWithSubmodule` is defined but never used in the test files. The actual submodule test (`TestRepositoryWithSubmodule`) uses `SetupEmptyRepo` instead and manually sets up submodules.

**Impact**: Reduces code reusability and may indicate incomplete test coverage or forgotten refactoring.

File: environment/filesystem.go [hunk 4]
```diff
@@ -183,3 +199,80 @@ func getMatchContext(contents string, matchIndex int) string {
...
+		}
+	}
+
+	cleanFilePath := filepath.Clean(filePath)
+
+	for _, submodulePath := range submodulePaths {
+		cleanSubmodulePath := filepath.Clean(submodulePath)
+
+		// Check if the file is exactly the submodule path or within it
+		if cleanFilePath == cleanSubmodulePath {
```
Comment (DaggerAssistant) [inline environment/filesystem.go:222]: 🐛 **Bug (Low):** **Edge case in path matching**: The submodule path validation logic checks for exact path match but may not handle edge cases correctly. If `cleanFilePath == cleanSubmodulePath`, it returns true, but this could match unintended paths.

**Example**: If submodule path is `sub` and file path is `sub-folder/file.txt`, the prefix check on line 261 would correctly reject it, but there might be edge cases with relative path resolution.

**Recommendation**: Add explicit test cases for edge cases like `../`, `./`, and similar paths in path matching logic.

**Review by DaggerAssistant — COMMENTED** [review]
## 🤖 Code Review

### Summary

This PR adds basic support for Git submodules in container environments. It detects submodule paths during environment creation, initializes them using host credentials, and enforces read-only access to submodule files within the container to prevent accidental modifications. The implementation caches submodule paths in environment state and validates file operations to block writes to submodule directories.

### Assessment

⚠️ The PR implements a **reasonable approach** to submodule support with proper read-only enforcement. However, there are several **functional issues** that need addressing:

**✅ Strengths:**
- **Smart architecture**: Caches submodule paths at creation time to avoid repeated detection
- **Proper isolation**: Read-only enforcement prevents accidental submodule modifications
- **Graceful degradation**: Failed submodule initialization doesn't break environment creation
- **Good test coverage**: Comprehensive test validates both read/write behavior

**⚠️ Critical Issues:**
1. **Inconsistent submodule detection**: Two different implementations with subtle behavioral differences
2. **Command execution vulnerability**: Shell command injection risk in `detectSubmodulePaths`
3. **Logic errors**: Missing submodule initialization in second commit command, unused helper
4. **Path validation issues**: Edge cases in path matching logic

**🔧 Functional Impact:**
- Submodule detection may fail inconsistently between host and container
- Shell injection could be exploited with malicious submodule paths
- Test helper is defined but never used, reducing test coverage

The core concept is sound but needs refinement to ensure reliable operation.

**Comment by DaggerAssistant** [comment]
## 🔍 CI Failure Analysis

### Failed Checks Summary
**Lint Check** - Failed due to unused function `(*Environment).detectSubmodulePaths` in `environment/filesystem.go`

### Root Cause Analysis
## 🔍 Root Cause Analysis

The lint failure is caused by an **unused function** that was left behind during development:

**File:** `environment/filesystem.go`  
**Function:** `(*Environment).detectSubmodulePaths()` (lines 203-235)  
**Issue:** This function is defined but never called anywhere in the codebase

### What Happened
During the implementation of submodule support, the author initially created a `detectSubmodulePaths` method in the `Environment` struct, but later moved the submodule detection logic to the `Repository` struct (`getSubmodulePaths` in `repository/git.go`). The original unused function was not removed, triggering the golangci-lint `unused` linter.

### Evidence
- The `unused` linter specifically flagged: `func '(*Environment).detectSubmodulePaths' is unused`
- A search shows `detectSubmodulePaths` is only defined, never called
- The working implementation is `(*Repository).getSubmodulePaths()` in `repository/git.go:431` which **is** being used in `repository/repository.go:231`

### Recommended Fixes
## 🛠️ Recommended Fixes

### **CRITICAL: Remove Unused Function**

**File:** `environment/filesystem.go`  
**Action:** Delete lines 203-235

Remove the entire unused `detectSubmodulePaths` function:

```go
// DELETE THIS ENTIRE FUNCTION (lines 203-235)
// detectSubmodulePaths returns a slice of submodule paths relative to the workdir
// This is called once during environment creation to cache the paths in state
func (env *Environment) detectSubmodulePaths(ctx context.Context) []string {
	// ... entire function body ...
}
```

### **Verification Steps**
After making the fix:
1. Run `golangci-lint run` locally to confirm the issue is resolved
2. Ensure all tests still pass with `go test ./...`
3. Verify the actual submodule functionality works (it should, since the working implementation is in `repository/git.go`)

### **Alternative Solutions** (Not Recommended)
- Add a `//nolint:unused` comment, but removing dead code is cleaner
- Wire up the function somewhere, but the Repository-based implementation is already working correctly

### Priority Assessment
**Critical** ❌ - This failure blocks the PR from merging and is trivial to fix

The lint check enforces code quality standards and prevents unused code from accumulating in the codebase. This is a straightforward cleanup issue that should be addressed before merge.

---
*Analysis completed by DagReview CI Agent*

**Comment by DaggerAssistant** [comment]
## 🔍 CI Failure Analysis

### Failed Checks Summary
- **Lint Check** (golangci-lint): Failed due to unused function

### Root Cause Analysis
The linter found an unused function in `environment/filesystem.go` at line 205:
```
func (*Environment).detectSubmodulePaths is unused
```

Looking at the code changes, I can see what happened:

1. The submodule detection logic was implemented in `environment/filesystem.go` as `detectSubmodulePaths()` 
2. The implementation was later moved to `repository/git.go` as `getSubmodulePaths()` (which is actively used in line 231 of `repository/repository.go`)
3. The original unused function was left behind in `environment/filesystem.go`

### Recommended Fixes

**Remove the unused function** from `environment/filesystem.go` lines 203-235:

```go
// Remove this entire function (lines 203-235):
// detectSubmodulePaths returns a slice of submodule paths relative to the workdir
// This is called once during environment creation to cache the paths in state
func (env *Environment) detectSubmodulePaths(ctx context.Context) []string {
    // ... entire function body ...
}
```

The correct implementation is already working properly in `repository/git.go` as `getSubmodulePaths()`, so simply deleting the unused method will resolve the lint failure.

### Additional Notes
- ✅ **Test Check**: Passed - All tests are working correctly with the current implementation
- ✅ **Build Check**: Passed - Code compiles successfully
- ⚠️ **Lint Check**: Failed - Only due to this unused function cleanup needed

This is a simple cleanup issue and should be a quick fix!

File: repository/git.go (hunk 2 above)
Comment (cwlbraa) [reply repository/git.go:544]: dead code, deleted detectSubmodulePaths

File: environment/integration/helpers.go (hunk 3 above)
Comment (cwlbraa) [reply environment/integration/helpers.go]: deleted

**Comment by grouville** [comment]
taking a look rn 👀

File: environment/filesystem.go [hunk 5]
```diff
@@ -124,6 +135,11 @@ func (env *Environment) FileEdit(ctx context.Context, explanation, targetFile, s
 }
 
 func (env *Environment) FileDelete(ctx context.Context, explanation, targetFile string) error {
+	// Check if the file is within a submodule
+	if err := env.validateNotSubmoduleFile(targetFile); err != nil {
+		return err
+	}
+
```
Comment (cwlbraa) [inline environment/filesystem.go:142]: there might be a better way to do this with `update-index --skip-worktree` or `update-index --assumed-unchanged`...

[BUT those seem to only work with individual files](https://www.reddit.com/r/git/comments/17xeppy/comment/k9n9ob6/) so maybe not.

File: environment/filesystem.go (hunk 4 above)
Comment (cwlbraa) [reply environment/filesystem.go:222]: i think cleaning the path covers this
